test(menu): refactor for cypress tests

stephenogorman opened the pull request:
### Proposed behaviour

Refactor Cypress tests for the `Menu` component to use new cypress-component-react framework for testing.
Add `menu.test.js`. 

### Current behaviour

Current process is the regular Cypress approach with BDD.

### Checklist

<!-- Each PR should include the following -->

- [x] Commits follow our style guide
- [ ] Related issues linked in commit messages if required
- [ ] Screenshots are included in the PR if useful
- [ ] All themes are supported if required
- [ ] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [ ] Storybook added or updated if required
- [ ] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [ ] Typescript `d.ts` file added or updated if required

#### QA

- [x] Tested in CodeSandbox/storybook
- [x] Add new Cypress test coverage if required
- [x] Carbon implementation matches Design System/designs
- [x] UI Tests GitHub check reviewed if required

### Additional context

<!-- Add any other context or links about the pull request here. -->

### Testing instructions

- [x] Run `npx cypress open --component` to check if there is newly added test.file
- [x] Check if the `menu.test.js` file passed
- [x] Check all tests were moved/refactored from .feature file -> .test.js file
- [x] Check all unnecessary files were deleted, including .json file
- [x] Run `npx cypress run --component` to check none of the other *.test.js files have regressed
- [x] Run `npx cypress run --e2e` to check none of the feature files have regressed

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+        getComponent("menu-fullscreen").should(
+          "have.css",
+          "left",
+          `${after}px`
```
do we need this `after` if it is still `0` for both examples?

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+    it("should verify that the width of the whole submenu with a long-labelled item is determined by the width of the item", () => {
+      CypressMountWithProviders(
+        <Box mb={150}>
+          {["white"].map((menuType) => (
```
do we need a `.map` here if we are using only one item in array?

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+    it("should verify that the width of the whole submenu with a long title is determined by the width of the title", () => {
+      CypressMountWithProviders(
+        <Box mb={150}>
+          {["white"].map((menuType) => (
```
same question here about the `.map`

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+        });
+    });
+
+    it.each([
```
we can wrap this test into a separate `describe` and pass what exactly `type` we are passing, as I did in line 104 in `pages.test.js` file

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+      ["30px", 30],
+      ["50px", 50],
+    ])(
+      "should verify Menu height is set by height prop when passed as an integer or a string",
```
same here for a type, we can actually wrap all these tests where we are passing this sentence `when passed as an integer or a string`

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+        <MenuComponent flex="auto" flexShrink={value} />
+      );
+
+      menu("box").should("have.css", "flex-shrink", shrinkText);
```
why here we have `menu("box")` but everywhere else just `menu()` ?

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+      (size, height, margin, width) => {
+        CypressMountWithProviders(
+          <Box mb={150}>
+            {["white"].map((menuType) => (
```
same here for the `.map(...)`

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+      (variant, color) => {
+        CypressMountWithProviders(
+          <Box mb={150}>
+            {["white"].map((menuType) => (
```
same here for the `.map(...)`

DlgSHi replied:
not fixed

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+        .click()
+        .then(() => {
+          // eslint-disable-next-line no-unused-expressions
+          expect(callback).to.have.been.called;
```
```suggestion
          expect(callback).to.have.been.calledOnce;
```

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+          .trigger("keydown", keyCode(key))
+          .then(() => {
+            // eslint-disable-next-line no-unused-expressions
+            expect(callback).to.have.been.called;
```
```suggestion
          expect(callback).to.have.been.calledOnce;
```

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+        .click({ multiple: true })
+        .then(() => {
+          // eslint-disable-next-line no-unused-expressions
+          expect(callback).to.have.been.called;
```
```suggestion
          expect(callback).to.have.been.calledOnce;
```

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+        .click()
+        .then(() => {
+          // eslint-disable-next-line no-unused-expressions
+          expect(callback).to.have.been.called;
```
```suggestion
          expect(callback).to.have.been.calledOnce;
```

DlgSHi reviewed (changes requested):
I think those TC are missing:
1. Scenario: Check that inner Menu without link is NOT available with tabbing in Fullscreen Menu
    Given I open "Menu Test" component page "menu-full-screen-story"
    When I continue to hit Tab key 8 times
    Then inner menu without active redirection is not focused

2. Scenario: If there is a Menu Item with a very long label inside a submenu, check that the width of the whole submenu is determined by this submenu item
    Given I open "Menu Test" component page "long-labels-story"
    When I hover over "first" expandable Menu component
    Then width of submenu should equal the width of the "second" submenu item

3. Scenario: If there is a Menu Item with a submenu that has a very long label, check the width of the whole submenu is determined by this parent item
    Given I open "Menu Test" component page "long-labels-story"
    When I hover over "second" expandable Menu component
    Then width of submenu should equal the width of "second" expandable Menu component

4. And few props are missing as well:
 - href, 
 - rel, 
 - parent, 
 - parentVariant

ZhuoyuJin commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+      cy.wait(50);
+      cy.focused().trigger("keydown", keyCode("uparrow"));
+      searchDefaultInput().should("have.focus");
+    });
```
Can we combine this test case with the above test case using it.each["downarrow", "uparrow"]?

ZhuoyuJin commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+};
+
+context("Testing Menu component", () => {
+  describe("should render Menu component", () => {
```
could you please change it to something like "check props for Menu component"

stephenogorman commented on src/components/menu/menu.test.js:
```diff
@@ -0,0 +1,1260 @@
+      (value, boolVal, state) => {
+        CypressMountWithProviders(<MenuComponentFullScreen isOpen={boolVal} />);
+
+        cy.viewport(1200, 800);
```
Test will fail when boolVal is true if we don't have the viewport.

stephenogorman commented:
> I think those TC are missing:
> 
>     1. Scenario: Check that inner Menu without link is NOT available with tabbing in Fullscreen Menu
>        Given I open "Menu Test" component page "menu-full-screen-story"
>        When I continue to hit Tab key 8 times
>        Then inner menu without active redirection is not focused
> 
>     2. Scenario: If there is a Menu Item with a very long label inside a submenu, check that the width of the whole submenu is determined by this submenu item
>        Given I open "Menu Test" component page "long-labels-story"
>        When I hover over "first" expandable Menu component
>        Then width of submenu should equal the width of the "second" submenu item
> 
>     3. Scenario: If there is a Menu Item with a submenu that has a very long label, check the width of the whole submenu is determined by this parent item
>        Given I open "Menu Test" component page "long-labels-story"
>        When I hover over "second" expandable Menu component
>        Then width of submenu should equal the width of "second" expandable Menu component
> 
>     4. And few props are missing as well:
> 
> 
>     * href,
> 
>     * rel,
> 
>     * parent,
> 
>     * parentVariant

I added tests for scenario 1 and the props which mere missing. I think scenarios 2 & 3 were actually covered, but the titles weren't clear enough. I have changed those now to:
"should verify if there is a Menu Item with a very long label inside a submenu, check that the width of the whole submenu is determined by this submenu item"
&
"should verify if there is a Menu Item with a submenu that has a very long label, check the width of the whole submenu is determined by this parent item"

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -277,35 +278,24 @@ context("Testing Menu component", () => {
-        "data-component",
-        "link"
-      );
+      for (let i = 0; i < 4; i++) {
```
```suggestion
      for (let i = 0; i < position.length; i++) {
```

DlgSHi commented on src/components/menu/menu.test.js:
```diff
@@ -1166,6 +996,216 @@ context("Testing Menu component", () => {
+      cy.viewport(1200, 800);
+    });
+
+    const menuClick = (position) => {
```
why we can't move this into `beforeEach` as well?

carbonci commented:
:tada: This PR is included in version 111.8.3 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/carbon-react/v/111.8.3)
- [GitHub release](https://github.com/Sage/carbon/releases/tag/v111.8.3)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
